### PR TITLE
New flag: --inline-callback obviates load.bzl

### LINKED ISSUE #60: Allow user to avoid creating `declare_maven`
I'm guessing it allows Stripe to customize the callback?
I'm not sure it's useful for most users. Perhaps allow users to tell bazel-deps "create a callback inside `3rdparty/workspace.bzl` while you write it?

### PR DESCRIPTION
**Pull request by cgrushko**
When turned on, a user doesn't have to provide a callback when calling maven_dependencies() from their WORKSPACE file.
Instead, maven_dependencies() will call a function defined in the same .bzl file that calls native.maven_jar() or equivalent.

The goal is to simplify setting up a project to work with bazel-deps.

Fixes #60 .

### CODE REVIEW

**Comment by cgrushko** [comment]
@johnynek 

There are existing tests for this, and I don't know Scala well enough (at all, actually) and my IntelliJ doesn't fully work with Scala. In short, if you add a skeleton test for Writer.scala, I'm happy to add tests for this new functionality.

The goal of this PR (and others I'll send) is to simplify setting up a project to work with bazel-deps. Ideally, a user could run bazel-deps on a pom.xml file and be done.

**Comment by cgrushko** [comment]
Alternatively, does it make more sense to put this option in the YAML file?

**Comment by johnynek** [comment]
This is great, and I'll merge, but I do agree with you that it should be an option in the yaml file. I think adding a global option there is the right way to go.

As for testing, we don't have a great story currently to test generation. We should ideally generate a workspace and then try to build it, but we haven't set that up yet and I don't want to block you.

So, if you see this PR: https://github.com/johnynek/bazel-deps/pull/66 it has an example.

**Comment by johnynek** [comment]
actually, making it a yaml option is the right call because we already have a maven server in the yaml file which we use to walk the dependencies. If we have this option there, you can generate with the same server, which makes a lot of sense.

**Comment by cgrushko** [comment]
Can you elaborate? I thought you already said it should be a yaml option?
How does that relate to the server?

**Comment by johnynek** [comment]
our code looks like:

```
def maven_load(hash):
    native.maven_jar(
        name = hash["name"],
        artifact = hash["artifact"],
        sha1 = hash["sha1"],
        repository = _nexus,
    )
    native.bind(
        name = hash["bind"],
        actual = hash["actual"],
    )
```

but actually we just want the one in `hash["repository"]` I guess...

Sorry, I am a bit foggy on some of this since it has been so long since some of this was written and I have not touched it.

So, take away: carry on. I think my comment was just confusing you. Putting the option in the yaml is still the plan I think we have agreed on.

**Comment by oscar-stripe** [comment]
actually, I don't know why we wouldn't make this the only option. If users want to ignore this function, that's fine. That will be the simplest to operate.

Here is the default, and if you want to customize, knock yourself out.

**Comment by cgrushko** [comment]
Do you mean generate it inline always?
I'm in favor ☺️

**Comment by johnynek** [comment]
yeah. Always generate inline. If you want to wire a custom callback, you can, but you really don't have to.

**Comment by cgrushko** [comment]
Updated, PTAL thanks!

File: src/scala/com/github/johnynek/bazel_deps/Writer.scala [hunk 1]
```diff
@@ -70,14 +70,28 @@ object Writer {
...
+        |        artifact = hash["artifact"],
+        |        sha1 = hash["sha1"],
+        |        repository = hash["repository"]
+        |    )
+        |    native.bind(
+        |        name = hash["bind"],
+        |        actual = hash["actual"]
+        |    )
+        |
+        |def maven_dependencies():
```
Comment (johnynek) [inline src/scala/com/github/johnynek/bazel_deps/Writer.scala]: this isn't quite what I meant, sorry.

Can you change this to:
```python
def maven_dependencies(callback = declare_maven):
```
so then people can override or not?

**Comment by johnynek** [comment]
you need to regenerate the 3rdparty for this repo. One of the tests is that running that tool produces no diff (to keep us honest about using the latest version of the code).

**Comment by cgrushko** [comment]
Ah, looks much better :)
Updated code, reran bazel-deps on itself, and updated docs.

**Pull request merged**